File: map-attribute-ios/MapAttributeApp/SettingsViewController.m
/*
 * Copyright (c) 2011-2022 HERE Europe B.V.
 * All rights reserved.
 */

#import "SettingsViewController.h"

@interface SettingsViewController ()

@property (nonatomic) NSArray* infos;
@property (nonatomic) NSArray* titles;

@property (nonatomic) NSDictionary* mapSchemes;
@property (nonatomic) NSArray* transitModes;

@property (nonatomic, weak) UISegmentedControl* mapSchemeCtrl;
@property (nonatomic, weak) UISegmentedControl* transitModeCtrl;
@property (nonatomic, weak) UISwitch* flowSwitch;
@property (nonatomic, weak) UISwitch* incidentSwitch;

@end

@implementation SettingsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.mapSchemes = @{
        @"NORMAL_DAY" : NMAMapSchemeNormalDay,
        @"HYBRID_DAY" : NMAMapSchemeHybridDay,
        @"TERRAIN_DAY" : NMAMapSchemeTerrainDay
    };
    self.transitModes = @[ @"NOTHING", @"STOP_AND_ACCESSE", @"EVERYTHING" ];
    self.titles = @[
        @"Map Schemes", @"Transit Mode Attributes", @"Traffic Layer Attributes"
    ];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView*)tableView
{

    return [self.titles count];
}

- (NSInteger)tableView:(UITableView*)tableView numberOfRowsInSection:(NSInteger)section
{

    if (section == 2)
        return 2;
    else
        return 1;
}

/**
 * Add segmetned switches for map scheme and transit mode, switch for traffic layer.
 **/
- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath
{

    UITableViewCell* cell;

    if (indexPath.section == 0)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"SegmentedCell"];
        if (cell == nil)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                          reuseIdentifier:@"SegmentedCell"];
        }
        UISegmentedControl* mapSchemeCtrl =
            [[UISegmentedControl alloc] initWithItems:[self.mapSchemes allKeys]];

        NSDictionary* textAttributes = @{NSFontAttributeName : [UIFont systemFontOfSize:14.0]};
        [mapSchemeCtrl setTitleTextAttributes:textAttributes forState:UIControlStateNormal];
        [mapSchemeCtrl setTitleTextAttributes:textAttributes forState:UIControlStateHighlighted];
        self.mapSchemeCtrl = mapSchemeCtrl;
        [cell.contentView addSubview:mapSchemeCtrl];

        NSInteger index = [[self.mapSchemes allValues] indexOfObject:_mapScheme];
        if (index != NSNotFound)
            mapSchemeCtrl.selectedSegmentIndex = index;

        mapSchemeCtrl.translatesAutoresizingMaskIntoConstraints = NO;
        NSDictionary* metrics = @{ @"ctrl" : mapSchemeCtrl };
        [cell.contentView
            addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(10)-[ctrl]-(10)-|"
                                                                   options:0
                                                                   metrics:nil
                                                                     views:metrics]];
        [cell.contentView
            addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-(10)-[ctrl]-(10)-|"
                                                                   options:0
                                                                   metrics:nil
                                                                     views:metrics]];
    }
    else if (indexPath.section == 1)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"SegmentedCell"];
        if (cell == nil)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                          reuseIdentifier:@"SegmentedCell"];
        }
        UISegmentedControl* transitModeCtrl =
            [[UISegmentedControl alloc] initWithItems:_transitModes];

        NSDictionary* textAttributes = @{NSFontAttributeName : [UIFont systemFontOfSize:12.0]};
        [transitModeCtrl setTitleTextAttributes:textAttributes forState:UIControlStateNormal];
        [transitModeCtrl setTitleTextAttributes:textAttributes forState:UIControlStateHighlighted];
        self.transitModeCtrl = transitModeCtrl;
        [cell.contentView addSubview:transitModeCtrl];

        transitModeCtrl.selectedSegmentIndex = (NSInteger)_transitDisplayMode;

        transitModeCtrl.translatesAutoresizingMaskIntoConstraints = NO;
        NSDictionary* metrics = @{ @"ctrl" : transitModeCtrl };
        [cell.contentView
            addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(10)-[ctrl]-(10)-|"
                                                                   options:0
                                                                   metrics:nil
                                                                     views:metrics]];
        [cell.contentView
            addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-(10)-[ctrl]-(10)-|"
                                                                   options:0
                                                                   metrics:nil
                                                                     views:metrics]];
    }
    else if (indexPath.section == 2)
    {
        cell = [tableView dequeueReusableCellWithIdentifier:@"SwitchCell"];

        if (cell == nil)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                          reuseIdentifier:@"SwitchCell"];
        }

        switch (indexPath.row)
        {
        case 0:
        {
            cell.textLabel.text = @"FLOW";
            UISwitch* flowSwitch = [[UISwitch alloc] init];
            cell.accessoryView = flowSwitch;
            flowSwitch.on = _trafficLayers & NMATrafficLayerFlow;
            self.flowSwitch = flowSwitch;
            break;
        }
        case 1:
        {
            cell.textLabel.text = @"INCIDENT";
            UISwitch* incidentSwitch = [[UISwitch alloc] init];
            cell.accessoryView = incidentSwitch;
            incidentSwitch.on = _trafficLayers & NMATrafficLayerIncidents;
            self.incidentSwitch = incidentSwitch;
            break;
        }
        default:
            break;
        }
    }
    return cell;
}

- (NSString*)tableView:(UITableView*)tableView titleForHeaderInSection:(NSInteger)section
{
    NSString* title;
    title = [self.titles objectAtIndex:section];
    return title;
}

#pragma mark - Table view delegate

- (CGFloat)tableView:(UITableView*)tableView heightForRowAtIndexPath:(NSIndexPath*)indexPath
{
    if (indexPath.section == 0 || indexPath.section == 1)
    {
        return 50.0;
    }
    return 40.0;
}

/**
 * It gets selection for map scheme
 **/
- (void)getMapSchemeSettings
{
    NSInteger segmentIndex = self.mapSchemeCtrl.selectedSegmentIndex;
    _mapScheme = [[self.mapSchemes allValues] objectAtIndex:segmentIndex];
}

/**
 * It gets selection for transit mode
 **/
- (void)getTransitModeSettings
{
    NSInteger segmentIndex = self.transitModeCtrl.selectedSegmentIndex;
    _transitDisplayMode = (NMAMapTransitDisplayMode)(segmentIndex);
}

/**
 * It gets selection for traffic layer
 **/
- (void)getTrafficLayersSettings
{
    _trafficLayers = 0;
    if (self.flowSwitch.isOn)
        _trafficLayers ^= NMATrafficLayerFlow;
    if (self.incidentSwitch.isOn)
        _trafficLayers ^= NMATrafficLayerIncidents;
}

/**
 * When cancel button is pressed, it calls delegates to perform.
 **/
- (IBAction)cancelBtnPressed:(id)sender
{
    [self.delegate settingsViewControllerDidCancel:self];
}

/**
 * When done button is clicked, it gets all selection for map attributes and
 *calls delegate to perform.
 **/
- (IBAction)doneBtPressed:(id)sender
{
    [self getMapSchemeSettings];
    [self getTransitModeSettings];
    [self getTrafficLayersSettings];
    [self.delegate settingsViewControllerDidDone:self];
}
@end

File: map-attribute-ios-swift/MapAttributeApp/SettingsViewController.swift
/*
 * Copyright (c) 2011-2020 HERE Europe B.V.
 * All rights reserved.
 */

import UIKit
import NMAKit

protocol SettingsViewControllerDelegate {
    func settingsViewControllerDidCancel(_ controller: SettingsViewController)
    func settingsViewControllerDidDone(_ controller: SettingsViewController)
}

class SettingsViewController: UITableViewController {

    var delegate: SettingsViewControllerDelegate?
    var mapScheme: String = NMAMapSchemeNormalDay
    var trafficLayers: Int = 0
    var transitDisplayMode: NMAMapTransitDisplayMode?

    let mapSchemes = [
        "NORMAL_DAY" : NMAMapSchemeNormalDay,
        "HYBRID_DAY" : NMAMapSchemeHybridDay,
        "TERRAIN_DAY" : NMAMapSchemeTerrainDay
    ]

    let titles = [
        "Map Schemes", "Transit Mode Attributes", "Traffic Layer Attributes"
    ]

    let transitModes = [
        "NOTHING", "STOP_AND_ACCESSE", "EVERYTHING"
    ]

    var mapSchemeCtrl: UISegmentedControl?
    var transitModeCtrl: UISegmentedControl?
    var flowSwitch: UISwitch = UISwitch()
    var incidentSwitch: UISwitch = UISwitch()

    override func numberOfSections(in tableView: UITableView) -> Int {
        return titles.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return section == 2 ? 2 : 1
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return titles[section]
    }

    /**
    * Add segmetned switches for map scheme and transit mode, switch for traffic layer.
    **/
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "SegmentedCell")

        guard let cell = tableView.dequeueReusableCell(withIdentifier: "SegmentedCell") else { return UITableViewCell() }

        if (indexPath.section == 0)
        {
            let mapSchemeCtrl = UISegmentedControl(items: Array(self.mapSchemes.keys))

            let textAttributes = [NSAttributedString.Key.font : UIFont.systemFont(ofSize: 14.0)]
            mapSchemeCtrl.setTitleTextAttributes(textAttributes, for:UIControl.State.normal)
            mapSchemeCtrl.setTitleTextAttributes(textAttributes, for:UIControl.State.highlighted)
            self.mapSchemeCtrl = mapSchemeCtrl
            cell.contentView.addSubview(mapSchemeCtrl)

            if let index = Array(self.mapSchemes.values).firstIndex(of: self.mapScheme) {
                mapSchemeCtrl.selectedSegmentIndex = index;
            }

            mapSchemeCtrl.translatesAutoresizingMaskIntoConstraints = false;
            let metrics = ["ctrl" : mapSchemeCtrl ]
            cell.contentView.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: "V:|-(10)-[ctrl]-(10)-|", metrics: nil, views: metrics)
            )
            cell.contentView.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: "H:|-(10)-[ctrl]-(10)-|", metrics: nil, views: metrics)
            )
        }
        else if (indexPath.section == 1)
        {
            let transitModeCtrl = UISegmentedControl(items: self.transitModes)

            let textAttributes = [NSAttributedString.Key.font : UIFont.systemFont(ofSize: 12.0)]
            transitModeCtrl.setTitleTextAttributes(textAttributes, for:UIControl.State.normal)
            transitModeCtrl.setTitleTextAttributes(textAttributes, for:UIControl.State.highlighted)
            self.transitModeCtrl = transitModeCtrl
            cell.contentView.addSubview(transitModeCtrl)

            if let index = self.transitDisplayMode?.rawValue {
                transitModeCtrl.selectedSegmentIndex = index
            }

            transitModeCtrl.translatesAutoresizingMaskIntoConstraints = false
            let metrics = ["ctrl" : transitModeCtrl ]
            cell.contentView.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: "V:|-(10)-[ctrl]-(10)-|", metrics: nil, views: metrics)
            )
            cell.contentView.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: "H:|-(10)-[ctrl]-(10)-|", metrics: nil, views: metrics)
            )
        }
        else if (indexPath.section == 2)
        {
            let switcher = UISwitch()
            cell.accessoryView = switcher;

            switch indexPath.row {
                case 0:
                    cell.textLabel?.text = "FLOW";
                    switcher.isOn = self.trafficLayers & NMATrafficLayer.flow.rawValue > 0
                    self.flowSwitch = switcher;
                case 1:
                    cell.textLabel?.text = "INCIDENT"
                    switcher.isOn = self.trafficLayers & NMATrafficLayer.incidents.rawValue > 0
                    self.incidentSwitch = switcher;
                default:
                    break;
            }
        }
        return cell;
    }

    @IBAction func done(_ sender: UIBarButtonItem) {
        // It gets selection for map scheme
        if let segmentIndex = self.mapSchemeCtrl?.selectedSegmentIndex {
            self.mapScheme = Array(self.mapSchemes.values)[segmentIndex]
        }
        // It gets selection for transit mode
        if let segmentIndex = self.transitModeCtrl?.selectedSegmentIndex {
            self.transitDisplayMode = NMAMapTransitDisplayMode(rawValue: segmentIndex)
        }
        // It gets selection for traffic layer
        self.trafficLayers = 0;
        if self.flowSwitch.isOn {
            self.trafficLayers ^= NMATrafficLayer.flow.rawValue;
        }
        if self.incidentSwitch.isOn {
            self.trafficLayers ^= NMATrafficLayer.incidents.rawValue;
        }

        self.delegate?.settingsViewControllerDidDone(self)
    }
    @IBAction func cancel(_ sender: UIBarButtonItem) {
        self.delegate?.settingsViewControllerDidCancel(self)
    }
}
